sigproc/src/OmnibusSigProc.cxx

```diff
@@ -931,9 +941,10 @@ void OmnibusSigProc::decon_2D_init(int plane){
     //// this is moved into wirecell.sigproc.main production of
     //// microboone-channel-responses-v1.json.bz2
     WireCell::Binning tbins(m_fft_nticks, cr_bins.min(), cr_bins.min() + m_fft_nticks*m_period);
-    Response::ColdElec ce(m_gain, m_shaping_time);
-    
-    const auto ewave = ce.generate(tbins);
+    //Response::ColdElec ce(m_gain, m_shaping_time);
+
+    //const auto ewave = ce.generate(tbins);
```

Comment:
It seems that the old code use "fft best length" as nbins, do we want to keep this feature?

Reply:
I'll comply with whatever people choose. Doesn't change much for standard ICARUS since the acquisition window is an "fft best length"

Reply:
I don't remember if in the meeting yesterday we came to a conclusive resolution on how we want to proceed with this. For now, I'll make sure that the "fft best length" is used hardcoded in both ColdElecResponse and WarmElecResponse.

Reply:
I think the consensus was to go with @lastgeorge (Xin's) suggestion to give the user a flag to choose between using the configured length or the "fft best length".  

If it isn't in CER/WER with a flag now, I don't think we need to get this exactly right in this particular PR.  We can un-hardcode it in some subsequent fix.

Reply:
> If it isn't in CER/WER with a flag now, I don't think we need to get this exactly right in this particular PR. We can un-hardcode it in some subsequent fix.

I think the "fft_flag" is turned off now in protoDUNE (see sp.jsonnet), also MicroBooNE is turned off, maybe @weihythu can double-check and confirm this.

Reply:
> I think the consensus was to go with @lastgeorge (Xin's) suggestion to give the user a flag to choose between using the configured length or the "fft best length".
> 
> If it isn't in CER/WER with a flag now, I don't think we need to get this exactly right in this particular PR. We can un-hardcode it in some subsequent fix.

I suggest we keep the `best_fft_length` function in OmnibusSigproc this time.

In OmnibusSigproc, `m_fft_nticks` has been handled here and should be fine: https://github.com/WireCell/wire-cell-toolkit/blob/f0ec58c8fa0b6df921861e0e28b37fb78e606e5b/sigproc/src/OmnibusSigProc.cxx#L733

While I am concerned about how the start point of the waveform is generated.  In WarmElecResponse (or ColdElecResponse), the start point of the waveform is configured by a parameter "start". And the original code reads a binning instance `tbins` to generate the waveform. To make them consistent, I think we have two options: 
1) pass the `tbins` to the function `IWaveform::waveform_samples()`
2) update the configuration in IWaveform (ColdElecRespone or WarmElecResponse), make sure the "start" of the waveform is consistent with the start point in 'tbins`.

I preferred option 1) because to maintain some consistency in the configuration sometimes is painful. What do you think? @ascarpel @brettviren

Reply:
I don't think option 1 is simply achievable, mostly because `waveform_samples()` returns a fully formed waveform. We can return a subset of it, but the array is filled by `configure`.  
I comment briefly your point 2): Brett can correct me if I am wrong here, but I am not very concerned by the start and ending point for mainly two reasons: 
1) both cold and warm electronics are defined in a quite short time domain and return 0 outside of it. 
2) I think the BINNING class wants a start and endpoint just to be more general, I bet it is used elsewhere too. I don't know why ElecResponse component wanted it to be user-configurable, but in principle, for how both electronics functions are defined it is a bit useless.

Reply:
I read the code again and realized we have two places that use `ColdElec`.  One is in `decon_2D_init`.

> WireCell::Binning tbins(m_fft_nticks, cr_bins.min(), cr_bins.min() + m_fft_nticks*m_period);
> //Response::ColdElec ce(m_gain, m_shaping_time);
> 
> //const auto ewave = ce.generate(tbins);
> auto ewave = (*m_elecresponse).waveform_samples();
> const WireCell::Waveform::compseq_t elec = Waveform::dft(ewave);

The other is in `init_overall_response`:

>   // since we only do FFT along time, no need to change dimension for wire ...
>   const size_t fine_nticks = fft_best_length(fravg.planes[0].paths[0].current.size());
>   int fine_nwires = fravg.planes[0].paths.size();
> 
>   WireCell::Waveform::compseq_t elec;
>   WireCell::Binning tbins(fine_nticks, 0, fine_nticks * fravg.period);
>   //Response::ColdElec ce(m_gain, m_shaping_time);
>   //auto ewave = ce.generate(tbins);
>   auto ewave = (*m_elecresponse).waveform_samples();

The first piece of code deals with the `PerChannelResponse`, and is actually only used in MicroBooNE now. The `cr_bins.min()`, i.e. t0, is -2750 in the data file `microboone-channel-responses-v1.json.bz2`. I think it is useful to let the ColdElecResponse know this t0, because the `dft()` returns not only amplitude but also phase information. @ascarpel 

The second piece of code explicitly calls `fft_best_length`. I think we need to let the ColdElecResponse know the length of ticks.

Reply:
Ok, thanks a lot for all these checks. I guess one can edit the function waveform to know the number of bins and fill the array there... I'll make a test. I am open to other ideas to test.

Reply:
So, I've been trying to change a bit the WarmElec response constructor to see if I can comply with what @wenqiang-gu suggests, but I can't think of any solution which doesn't involve changing ether the IConfigure or the IWaveform classes. This would be a major change in the code, perhaps @brettviren  can provide his opinion about it. One, less elegant alternative is just provide the OmnibusSigproc.cxx component with a flag that simply ask if the user whishes to use the warm electronics and if so just configure the functional form form Response.h. In this case we drop the external configuration part, but we save a bunch of headaches for now. I am not a bit fan of this option: in principle with the correct game of configuration files, one should always be able to set externally the precise requirements that are needed for the experiments, unelss something is internally changed and the user has no control of it.

Reply:
Sorry to be late to this thread.

I'm still a bit confused why `cr_bins` from the per-channel-response is used to define the domain of the `ColdElec` response waveform:

https://github.com/WireCell/wire-cell-toolkit/blob/f0ec58c8fa0b6df921861e0e28b37fb78e606e5b/sigproc/src/OmnibusSigProc.cxx#L933

These two responses are convolved and so any time shift each may have individually will be added together in the result.  So, I think either there is a double-counted shift being artificially created or there is some incorrect shift baked into the per-channel-response function and using `cr_bins` undoes that mistake.  Or, `cr_bins.min()` is in fact always 0 and so no bug shows up but only by accident.

It would be best to understand this.

But to get this PR done, I'm okay with going forward with the C++ level solution.  Here is what I suggested on our meeting:

1) In `IWaveform` **add** a new method

``` c++
virtual sequence_type waveform_samples(const WireCell::Binning& tbins) const = 0;
```
Two things to note: Leave the existing `waveform_samples()` so previous code continues to work.  Also, return-by-value instead of the return-by-const-reference used by the existing method.  It was a mistake to return by const reference and we do it in many interfaces.  One day we might fix that but for now we can at least do the right thing for new interface methods.  And, in particular because the return of this new method is a function of arbitrary input there is no sane way to provide the needed caching to return by const reference.  (Well, an LRU cache could be used, I guess, but I don't think this will help anything and just add code complication.)

2) Implement this new method in all existing `IWaveform` subclasses.  Right now I see just three:

```
$ grep IWaveform */inc/*/*.h|grep public
gen/inc/WireCellGen/ElecResponse.h:        class ElecResponse : public IWaveform, public IConfigurable {
gen/inc/WireCellGen/RCResponse.h:        class RCResponse : public IWaveform, public IConfigurable {
gen/inc/WireCellGen/ResponseSys.h:        class ResponseSys : public IWaveform, public IConfigurable {
```

My copy of the code doesn't reflect the new implementations in this PR so I guess `ElecResponse` is now replaced with the CER and WER versions.

3) Continue with replacing the hard-wired `ColdElec` with an `IWaveform`.

Let me know if this isn't clear or doesn't seem correct.  Thanks!

Reply:
Sounds perfect to me! I believe there was an offset in the data file, so people want to double-correct this.

Reply:
@brettviren, @wenqiang-gu, I have implemented the new method to IWaveform and updated the corresponding components. I tested for ICARUS only. Please, check if this new solution can be satisfactory. I have touched the following: 
```
gen/inc/WireCellGen/WarmElecResponse.h
gen/inc/WireCellGen/ColdElecResponse.h
gen/inc/WireCellGen/RCResponse.h       
gen/inc/WireCellGen/ResponseSys.h        
```